### ISSUE TITLE
compiling libhybris error with arch=x86

### ISSUE THREAD

**Issue by dliang0406** [libhybris/libhybris#282]
Hello,

I want to run libhybris on my x86 device. However after I ran "./autogen.sh --prefix=$HOME/hybris_rootfs --with-android-headers=$ANDROID_HEADERS --enable-arch=x86 --enable-wayland" and "make", I got the following error:
dlfcn.c:203:7: error: initializer element is not constant
       st_value: (Elf32_Addr) &android_dlopen,
       ^
It seems the C99 spec doesn't allow initializing a global varible with a non constant value.

I'm confused as it seems I'm the only person who met such problem. I can't find the solution anywhere through the internet. Or maybe this issue is too simple to be an issue. But with my limited compiler knowledge, I can't figure out the way to walk around it.

Please Help!!! @stskeeps

**Comment by morphis** [comment]
@dliang0406 Which compiler are you using?

**Comment by dliang0406** [comment]
The host gcc.

**Comment by morphis** [comment]
Which version?

**Comment by dliang0406** [comment]
gcc 4.8.4

**Comment by dliang0406** [comment]
My host is Ubuntu14.04

**Comment by dliang0406** [comment]
I downgraded my host gcc to 4.7.3 and still got the same error.

However, I tried the arm-linux-gnueabi-gcc(4.7.3) and everything looks fine.

**Comment by dliang0406** [comment]
Look like a GCC bug:
https://gcc.gnu.org/bugzilla/show_bug.cgi?id=38354

**Comment by dliang0406** [comment]
The compiling is okay after using g++ instead.

**Comment by ycavan** [comment]
@dliang0406 could you let me know what you modified to get hybris to compile with g++?  ie, did you make changes to the Makefile(s) or was it a configure switch? thx

**Comment by dliang0406** [comment]
@ycavan I just modifed the generated Makefile to change CC to g++ and add one CFLAGS -fpermissive.
Since g++ is used, symbols are C++ mangled styles. Other place using these functions should be modified to use the C++ symbols like _Z14android_dlopenPKci.

My problem is that my host gcc is a x86_64 version. Even I made such changes and the compilling is OK, the hybris linker cannot work. I'm prepareing a x86_64 patch to add the x86_64 support for the linker.

**Comment by ycavan** [comment]
@dliang0406 Please let me know when you have your patch ready.  Even after changing CC = gcc to CC = g++ -fpermissive I get the same initializer error.  ( GCC suite v 5.1.0 for x86_64 )

**Comment by dliang0406** [comment]
Not in a short time. I'm busy with sth else. What do you want to do?

**Comment by ycavan** [comment]
@dliang0406 I'm trying to get libhybris compiled for Arch Linux x86_64 chroot on my Zenfone 2 to make use of the Android powervr 6430 ( rogue ) gpu drivers.

**Comment by dliang0406** [comment]
@ycavan Zenfone2 doesn't use x86_64 in user space I think. You don't need to bother with the 64b issue.

I tried the gcc of i386, the initializer error is gone.

**Comment by ycavan** [comment]
@dliang0406 I'm running Arch x86_64 right now to build the kernel on the phone itself ( I prefer to build my kernels on the actual devices ).  I guess I could create a different chroot that isn't 64 bit and try it.

**Comment by dliang0406** [comment]
@ycavan The zenfone2 kernel is x86_64 arch. But I'm a little confused, I thought libhybris had nothing to do with kernel.

**Comment by ycavan** [comment]
@dliang0406 The zenfone2 kernel is actually AOSP_x86-eng even though the cpu is 64 bit.  AOSP requires an x86_64 environment to compile modern Android kernels; hence I'm using an x86_64 environment.  I was concerned because you said your environment was also x86_64 and the libhybris library wasn't working for you after it compiled? ( mine still won't compile even though I changed gcc to g++ w/ cflags= -fpermissive in the generated Makefile )

basically, if a pure i386 chroot will allow compilation & use of the library, I'll try to start up a different chroot that's i386 based.

**Comment by dliang0406** [comment]
@ycavan i386 rootfs can make the compilation work. I tried ubuntu core i386 version. It's OK. But I don't know how to hook this ubuntu core with my Android device. Any ideas?

**Comment by ycavan** [comment]
@dliang0406 I was thinking that we could create a symlink to /system/libs when we create the chroot?  I'm not familiar enough w/ libhybris and I was hoping to play with it after I got it to compile on my x86_64 chroot.

**Comment by dliang0406** [comment]
@ycavan I bet your location is in Asia as you replied me soooo quickly. :)

**Comment by ycavan** [comment]
@dliang0406 Iowa in the USA. :)

**Comment by dliang0406** [comment]
@ycavan you're a night owl. :)
I have to focus on the 32bit libhybris first.
I just change the CC to g++/-fpermissive in ./common/jb/Makefile. I don't touch elsewhere.

**Comment by ycavan** [comment]
After changing CC to g++ & CC flags to -fpermissive in ./common/jb/Makefile:

dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
 };
 ^
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:240:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:284:1: error: C99 designator ânameâ outside aggregate initializer
 };
 ^
dlfcn.c:284:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:284:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:284:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:284:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:284:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:284:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:284:1: sorry, unimplemented: non-trivial designated initializers not supported
dlfcn.c:284:1: sorry, unimplemented: non-trivial designated initializers not supported
Makefile:421: recipe for target 'libandroid_linker_la-dlfcn.lo' failed
make[3]: *** [libandroid_linker_la-dlfcn.lo] Error 1

**Comment by dliang0406** [comment]
@ycavan This is a different story...C++ spec requires the members in a data structure are initialized in order.

**Comment by ycavan** [comment]
huh.  so that structure is out of order & it compiled for you?

**Comment by dliang0406** [comment]
@ycavan corret. I reordered the members and add the missing one with zero initialized.

**Comment by ycavan** [comment]
gotcha.  I guess I'll have to go through that struct.

**Comment by ycavan** [comment]
@dliang0406 What did you end up using for soinfo libdl_info?  This is what's currently in the file ( one of the out of order ones ):

soinfo libdl_info = {
    name: "libdl.so",
    flags: FLAG_LINKED,

    strtab: ANDROID_LIBDL_STRTAB,
    symtab: libdl_symtab,

    refcount: 1,
    nbucket: 1,
#if defined(ANDROID_ARM_LINKER)
    nchain: 8,
#else
    nchain: 7,
#endif
    bucket: libdl_buckets,
    chain: libdl_chains,
};

**Comment by dliang0406** [comment]
soinfo libdl_info = {
    "libdl.so",
    phdr: 0,
    phnum: 0,
    entry: 0,
    base: 0,
    size: 0,
    unused: 0,
    dynamic: 0,
    wrprotect_start: 0,
    wrprotect_end: 0,
    next: 0,
    flags: FLAG_LINKED,

    strtab: ANDROID_LIBDL_STRTAB,
    symtab: libdl_symtab,

    nbucket: 1,
    nchain: 7,
    bucket: libdl_buckets,
    chain: libdl_chains,
};

**Comment by ycavan** [comment]
I'm not doing something right.  would you be able to post your dlfcn.c file somewhere?

**Comment by dliang0406** [comment]
@ycavan do you give me email address?

**Comment by ycavan** [comment]
@dliang0406 ycavan@yahoo.com & thx :)

**Comment by vicamo** [comment]
Tried to compile with following command with success. Probably a non-issue now.
```
for level in 14 15 16 17 18 19 21 22 23; do
  ./hybris/autogen.sh --prefix=/usr \
    --with-default-egl-platform=hwcomposer \
    --enable-debug --enable-trace \
    --with-default-hybris-ld-library-path=/system/lib:/vendor/lib \
    --with-android-headers=/home/vicamo/work/android/aosp/platform/headers/android-${level} \
    --enable-arch=x86 \
  && make -j 16 \
  && git clean -d -f -x;
  [ x$? == x0 ] || break;
done
```
